build-support/src/main/kotlin/com/jakewharton/mosaic/buildsupport/MosaicBuildExtensionImpl.kt

```diff
@@ -0,0 +1,154 @@
+
+				val name = target.name + "Test"
+				val nameUpper = name.replaceFirstChar(Char::uppercase)
+
+				val mainJarProvider = project.tasks.named(target.artifactsTaskName)
+
+				val testCompilation = target.compilations.named(TEST_COMPILATION_NAME)
+				val testClassesProvider = testCompilation.map { it.output.allOutputs }
+				val testDependenciesProvider = testCompilation.map {
+					it.runtimeDependencyFiles?.filter { it.isFile }
+						?: FileCollectionFactory.empty()
+				}
+
+				val testJarProvider = project.tasks.register("jar$nameUpper", Jar::class.java) {
+					it.from(testClassesProvider)
+					it.archiveAppendix.set(target.name)
+					it.archiveClassifier.set("tests")
+				}
+
+				val testScriptsProvider = project.tasks.register("scripts$nameUpper", CreateStartScripts::class.java) {
+					it.outputDir = project.layout.buildDirectory.dir("scripts/$name").get().asFile
+					it.applicationName = base.archivesName.get() + "-test"
+
+					// The classpath property is not lazy, so we need explicit dependencies here.
+					it.dependsOn(mainJarProvider)
+					it.dependsOn(testJarProvider)
+					it.dependsOn(testDependenciesProvider)
+					// However, this 'plus' result will be live, and can still be set at configuration time.
+					val classpath = mainJarProvider.get().outputs.files
+						.plus(testJarProvider.get().outputs.files)
+						.plus(testDependenciesProvider.get())
+					it.classpath = classpath
+
+					it.mainClass.set(
+						testClassesProvider.zip(testDependenciesProvider) { testClasses, testDependencies ->
+							val testFqcns = gradleSupport.detectTestClassNames(
+								testClasses.asFileTree,
+								testClasses.files.toList(),
+								testDependencies.files.toList()
+							)
+							"org.junit.runner.JUnitCore ${testFqcns.joinToString(" ") { """"$it"""" }}"
+						}
+					)
+				}
+
+				val installProvider = project.tasks.register("install${nameUpper}Distribution", Copy::class.java) {
+					it.group = "distribution"
+					it.description = "Installs $name as a distribution as-is."
+
+					it.into("bin") {
+						it.from(testScriptsProvider)
+					}
+					it.into("lib") {
+						it.from(testJarProvider)
+						it.from(mainJarProvider)
+						it.from(testDependenciesProvider)
+					}
+					it.destinationDir = project.layout.buildDirectory.dir("install/$name").get().asFile
+				}
+
+				project.tasks.register("zip${nameUpper}Distribution", Zip::class.java) {
+					it.group = "distribution"
+					it.description = "Bundles $name as a distribution."
+
+					it.from(installProvider)
+					it.destinationDirectory.set(project.layout.buildDirectory.dir("dist"))
+					it.archiveAppendix.set(target.name)
+					it.archiveClassifier.set("tests")
+				}
+
+				// TODO add to archives?
+			}
+		}
+	}
+
+	interface GradleSupport {
+		fun detectTestClassNames(
+			testClasses: FileTree,
+			testClassDirectories: List<File>,
+			testClasspath: List<File>,
+		): List<String>
+	}
+
+	class Gradle_8_10_Support : GradleSupport {
```

Comment:
We're on Gradle 8.12 now, but when I originally wrote this it worked on Gradle 8.10...